lib/master/worker_manager.go

```diff
@@ -0,0 +1,494 @@
+	CallbackWithError = func(ctx context.Context, handle WorkerHandle, err error) error
+)
+
+type WorkerManager struct {
+	mu            sync.Mutex
+	workerEntries map[libModel.WorkerID]*workerEntry
+	state         workerManagerState
+
+	workerMetaClient *metadata.WorkerMetadataClient
+	messageSender    p2p.MessageSender
+
+	masterID libModel.MasterID
+	epoch    libModel.Epoch
+
+	onWorkerOnlined       Callback
+	onWorkerOfflined      CallbackWithError
+	onWorkerStatusUpdated Callback
+	onWorkerDispatched    CallbackWithError
+
+	eventQueue chan *masterEvent
+	closeCh    chan struct{}
+	errCh      chan error
+
+	clock clock.Clock
+
+	timeouts config.TimeoutConfig
+
+	wg sync.WaitGroup
+}
+
+type workerManagerState int32
+
+const (
+	workerManagerReady = workerManagerState(iota + 1)
+	workerManagerWaitingHeartbeat
+)
+
+func NewWorkerManager(
+	masterID libModel.MasterID,
+	epoch libModel.Epoch,
+	meta metaclient.KVClient,
+	messageSender p2p.MessageSender,
+	onWorkerOnline Callback,
+	onWorkerOffline CallbackWithError,
+	onWorkerStatusUpdated Callback,
+	onWorkerDispatched CallbackWithError,
+	isInit bool,
+	timeoutConfig config.TimeoutConfig,
+	clock clock.Clock,
+) *WorkerManager {
+	state := workerManagerReady
+	if !isInit {
+		state = workerManagerWaitingHeartbeat
+	}
+
+	ret := &WorkerManager{
+		workerEntries: make(map[libModel.WorkerID]*workerEntry),
+		state:         state,
+
+		workerMetaClient: metadata.NewWorkerMetadataClient(masterID, meta),
+		messageSender:    messageSender,
+
+		masterID: masterID,
+		epoch:    epoch,
+
+		onWorkerOnlined:       onWorkerOnline,
+		onWorkerOfflined:      onWorkerOffline,
+		onWorkerStatusUpdated: onWorkerStatusUpdated,
+		onWorkerDispatched:    onWorkerDispatched,
+
+		eventQueue: make(chan *masterEvent, 1024),
+		closeCh:    make(chan struct{}),
+		errCh:      make(chan error, 1),
+
+		clock:    clock,
+		timeouts: timeoutConfig,
+	}
+
+	ret.wg.Add(1)
+	go func() {
+		defer ret.wg.Done()
+		if err := ret.runBackgroundChecker(); err != nil {
+			select {
+			case ret.errCh <- err:
+				log.L().Warn("runBackgroundChecker encountered error",
+					zap.String("master-id", masterID),
+					zap.Error(err))
+			default:
+				log.L().Warn("runBackgroundChecker error dropped",
+					zap.String("master-id", masterID),
+					zap.Error(err))
+			}
+		}
+	}()
+
+	return ret
+}
+
+func (m *WorkerManager) Close() {
+	close(m.closeCh)
+	m.wg.Wait()
+}
+
+// InitAfterRecover should be called after the master has failed over.
+// This method will block until a timeout period for heartbeats has passed.
+func (m *WorkerManager) InitAfterRecover(ctx context.Context) error {
+	timeoutInterval := m.timeouts.WorkerTimeoutDuration + m.timeouts.WorkerTimeoutGracefulDuration
+	timer := m.clock.Timer(timeoutInterval)
+	defer timer.Stop()
+
+	select {
+	case <-ctx.Done():
+		return errors.Trace(ctx.Err())
+	case <-timer.C:
+		// Wait for the worker timeout to expire, so
+		// that we maintain all the workers that are alive.
+	}
```

Comment:
Does this mean master won't start to work until `m.timeouts.WorkerTimeoutDuration + m.timeouts.WorkerTimeoutGracefulDuration` (20s) passes

Reply:
Yes. We can change it so that if all workers recorded in meta have sent us a heartbeat, the master resumes working. But in cases where there is at least one worker crashing, the master still needs to wait for `20s`.